design-documents/distributed-deployment.md

```diff
@@ -0,0 +1,217 @@
+    * How to manage obsolete dependencies (you remove module on once instance, need to remove relevant parts on another)?
+    
+    ![Module dependencies on services](distributed-deployment/module-dependencies-on-services.png)
+
+2. Distribute different projects and introduce new type of configuration for declaring dependencies of packages on instances/services
+
+    Modules would define on which instances they need to be installed and tool would modify composer.json files of these instances to add dependencies to require section.
+    
+    CatalogExtensionAdminUi/composer.json
+    ```
+    {
+        "name": "vendor/catalog-extension-admin-ui",
+        ...
+        "instances": [
+            "admin-ui"
+        ]
+    }
+    ```
+    
+    As when we install extension we use metapackage name (metapackage includes all parts of the extension: AdminUi, Ui, Cron, WebApi) and metapackage is not being installed or saved anywhere, we need to add this information to the composer.json on the instance.
+    
+    admin-ui-instance/composer.json
+    ```
+    {
+        "name": "magento/admin-ui-project",
+        "type": "project",
+        "require": {
+            "vendor/catalog-extension-admin-ui": "*"
+        },
+        "installed-packages": [
+            "vendor/catalog-extension"
+        ]
+    }
+    ```
+    
+    This would allow to understand which packages that are installed on the instance are part of extension we want to remove and remove these packages from all instances.
+    
+    Pros
+     * The same mechanism can be used for distributed deployment of services
+    
+    Cons
+    * Will require us to publish more projects
+    * Need to add new type of configuration
+    
+    Examples of using command
+    
+    ```
+    instance-manager install --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager uninstall --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager update --instances "./admin-ui ./ui ./cron ./webapi"
+    ```
+    
+    Some of corner cases tool need to support
+    1. When upgrading extension that shares common package with another extension, make sure that version of the common package satisfies metapackage of another package
+    2. When removing package from one of the instances that is part of the extension user want to remove, check that package is not part of any other installed extension
+    3. When updating packages, check that new versions of the packages satisfy version requirements of metapackage
+    4. When updating extension find new versions of extensions that consist from packages that have their dependency/versions met (dependencies on Magento packages). When calculating versions, we probably will have to take into account Magento package versions on other instances
+    
+    #### Tool architecture
```

Comment:
As we can make the only sync upgrades, I think it doesn't make sense to introduce a specific tool to do it. Composer should care about dependencies resolving and the tool might be used to cover cases which the composer doesn't resolve.